pkg/storage/utils/decomposedfs/upload.go

```diff
@@ -383,12 +397,60 @@ func (fs *Decomposedfs) GetUpload(ctx context.Context, id string) (tusd.Upload,
 	return &fileUpload{
 		info:     info,
 		binPath:  info.Storage["BinPath"],
-		infoPath: infoPath,
+		infoPath: filepath.Join(fs.o.Root, "uploads", id+".info"),
 		fs:       fs,
 		ctx:      ctx,
 	}, nil
 }
 
+// ListUploads returns a list of all incomplete uploads
+func (fs *Decomposedfs) ListUploads() ([]tusd.FileInfo, error) {
+	return fs.uploadInfos()
+}
+
+func (fs *Decomposedfs) uploadInfos() ([]tusd.FileInfo, error) {
+	infos := []tusd.FileInfo{}
+	infoFiles, err := filepath.Glob(filepath.Join(fs.o.Root, "uploads", "*.info"))
+	if err != nil {
+		return nil, err
+	}
+
+	idRegexp := regexp.MustCompile(".*/([^/]+).info")
+	for _, info := range infoFiles {
+		match := idRegexp.FindStringSubmatch(info)
+		if match == nil || len(match) < 2 {
+			continue
+		}
+		info, err := fs.readInfo(match[1])
+		if err != nil {
+			return nil, err
+		}
+		infos = append(infos, info)
+	}
+	return infos, nil
+}
+
+// PurgeExpiredUploads scans the fs for expired downloads and removes any leftovers
+func (fs *Decomposedfs) PurgeExpiredUploads(purgedChan chan<- tusd.FileInfo) error {
+	infos, err := fs.uploadInfos()
+	if err != nil {
+		return err
+	}
+
+	for _, info := range infos {
+		expires, err := strconv.Atoi(info.MetaData["expires"])
+		if err != nil {
+			continue
+		}
+		if int64(expires) < time.Now().Unix() {
+			purgedChan <- info
+			os.Remove(info.Storage["BinPath"])
+			os.Remove(filepath.Join(fs.o.Root, "uploads", info.ID+".info"))
```

Comment:
Should we log errrors here? We might want to inform "someone" when there is a problem removing files from the disk..

Reply:
:+1: sounds reasonable, I'll add logging for those cases and merge. And sure, let me know when you need help with merging.

Reply:
I took a look and getting a logging facility in there would be quite involved. I think it's fine to just error out if something goes wrong instead. Errors during deletion aren't expected anyway and should be investigated if they happen. Let me know if that doesn't work for you.